docs/proposals/20210210-insulate-users-from-kubeadm-API-changes.md

```diff
@@ -0,0 +1,226 @@
+---
+title: Insulate users from kubeadm API version changes
+authors:
+- "@fabriziopandini"
+reviewers:
+- "@vincepri"
+creation-date: 2021-02-10
+last-updated: 2021-02-10
+status: implementable
+see-also:
+- "/docs/proposals/20190610-machine-states-preboot-bootstrapping.md"
+replaces:
+superseded-by:
+---
+
+# Insulate users from kubeadm API version changes
+
+## Table of Contents
+
+[Tools for generating](https://github.com/ekalinin/github-markdown-toc)
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+## Summary
+
+Make CABPK and KCP to use more recent versions of the kubeadm API and insulate users from
+kubeadm API version changes.
+
+## Motivation
+
+Cluster bootstrap provider for kubeadm (CABPK) and the control plane provider for kubeadm
+(KCP) API still relying on kubeadm v1beta1 API, which has been deprecated, and it is going
+to be removed ASAP.
+
+While moving to a more recent version of the kubeadm API is a required move, Cluster API
+should take this opportunity stop to exposing the kubeadm API types in the
+KubeadmConfig/KubeadmControlPlane specs, and instead use our own types.
+
+This would allow to separate what users fill in in the KubeadmConfig/KubeadmControlPlane
+from which kubeadm API version Cluster API end up using in the bootstrap data.
+
+### Goals
+
+- Define a stop-gap for using most recent version of the kubeadm API in Cluster API
+  v1alpha3 - without introdicing any breaking change -.
+- Define how to stop to exposing the kubeadm v1betax types in the KubeadmConfig/KubeadmControlPlane
+  specs for v1alpha4.
+- Define how to use the right version of the kubeadm types generating our kubeadm yaml
+  file and when interacting with the kubeadm-config ConfigMap.
+- Ensure a clean and smooth v1alpha3 to v1aplha4 upgrade experience.
+
+### Non-Goals
+
+- Adding or removing fields in KubeadmConfig/KubeadmControlPlane spec.
+- Introduce behevioral changes in CABPK or KCP.
+
+### Future work
+
+- Evaluate improvements for the Cluster API owned version of KubeadmConfig/KubeadmControlPlane
+  specs types.
+- Make it possible for the Cluster API users to take benefit of changes introduced
+  in recent versions of the Kubeadm API.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a user, I want to user kubeadm providers for Cluster API (CABPK & KPC) in the same
+way across the entire spectrum of supported kubernetes versions.
+
+#### Story 2
+
+As the kubeadm tool, I want the kubeadm providers for Cluster API (CABPK & KPC) to
+use the latest kubeadm API supported by the target kubeadm/kubernetes version.
+
+### Requirements
+
+R1 - avoid breaking changes in v1alpha3
+R2 - ensure a clean v1alpha3 to v1alpha4 upgrade
+
+### Implementation Details/Notes/Constraints
```

Comment:
i think this section is missing the actual implementation details of how the "insulation" will happen?

Reply:
The insulation happens by implementing serialization/deserialization conversion on top of the kubeadm config types already defined in Cluster API.